apps/emqx/src/emqx_channel.erl

```diff
@@ -1303,66 +1308,65 @@ handle_timeout(
     case emqx_keepalive:check(StatVal, Keepalive) of
         {ok, NKeepalive} ->
             NChannel = Channel#channel{keepalive = NKeepalive},
-            {ok, reset_timer(alive_timer, NChannel)};
+            {ok, reset_timer(keepalive, NChannel)};
         {error, timeout} ->
             handle_out(disconnect, ?RC_KEEP_ALIVE_TIMEOUT, Channel)
     end;
 handle_timeout(
     _TRef,
-    retry_delivery,
+    Name,
     Channel = #channel{conn_state = disconnected}
-) ->
+) when ?IS_COMMON_SESSION_TIMER(Name) ->
     {ok, Channel};
 handle_timeout(
     _TRef,
-    retry_delivery,
+    Name,
     Channel = #channel{session = Session, clientinfo = ClientInfo}
-) ->
-    case emqx_session:retry(ClientInfo, Session) of
-        {ok, NSession} ->
+) when ?IS_COMMON_SESSION_TIMER(Name) ->
+    % NOTE
+    % Responsibility for these timers is smeared across both this module and the
+    % `emqx_session` module: the latter holds configured timer intervals, and is
+    % responsible for the actual timeout logic. Yet they are managed here, since
+    % they are kind of common to all session implementations.
+    case emqx_session:handle_timeout(ClientInfo, Name, Session) of
+        {ok, Publishes, NSession} ->
             NChannel = Channel#channel{session = NSession},
-            {ok, clean_timer(retry_timer, NChannel)};
+            handle_out(publish, Publishes, clean_timer(Name, NChannel));
         {ok, Publishes, Timeout, NSession} ->
             NChannel = Channel#channel{session = NSession},
-            handle_out(publish, Publishes, reset_timer(retry_timer, Timeout, NChannel))
+            handle_out(publish, Publishes, reset_timer(Name, Timeout, NChannel))
     end;
 handle_timeout(
     _TRef,
-    expire_awaiting_rel,
-    Channel = #channel{conn_state = disconnected}
-) ->
-    {ok, Channel};
-handle_timeout(
-    _TRef,
-    expire_awaiting_rel,
+    {emqx_session, Name},
     Channel = #channel{session = Session, clientinfo = ClientInfo}
 ) ->
-    case emqx_session:expire(ClientInfo, awaiting_rel, Session) of
-        {ok, NSession} ->
-            NChannel = Channel#channel{session = NSession},
-            {ok, clean_timer(await_timer, NChannel)};
-        {ok, Timeout, NSession} ->
-            NChannel = Channel#channel{session = NSession},
-            {ok, reset_timer(await_timer, Timeout, NChannel)}
+    case emqx_session:handle_timeout(ClientInfo, Name, Session) of
+        {ok, [], NSession} ->
+            {ok, Channel#channel{session = NSession}};
+        {ok, Replies, NSession} ->
+            handle_out(publish, Replies, Channel#channel{session = NSession})
```

Comment:
No, I think we don't. The point of `{emqx_session, _}` timers is that the session implementation is exclusively responsible for them.

Reply:
The changeset is stable (barring few good points and nits in the comments). Common session timers are handled in the previous clause. This is a plumbing for _custom_ timers to be used inside `emqx_persistent_session_ds` later. Admittedly, this is a kind of leftover from the previous iteration, do you think it would be less confusing if we move these into a followup PR?

Reply:
Thanks for the input! Backed them out as part of e3b5d75.